# Local Embeddings
In this Python notebook, we will try running embedding models locally and compare their performances.

## Using Llama-Index
We'll be making use of LlamaIndex to help us download the embedding models from Hugging Face.

In [3]:
# If you face issues running the code cell below, uncomment the 2 lines below and run this code cell to install some libraries

!pip install llama-index-llms-huggingface
!pip install llama-index-embeddings-huggingface

In [6]:
import os
# Optional, but set llamaindex cache dir to ../cache dir here. Default is system tmp
# This way, we can easily see downloaded artifacts
os.environ['LLAMA_INDEX_CACHE_DIR'] = os.path.join(os.path.abspath('../'), 'cache')

# from llama_index.embeddings import HuggingFaceEmbedding
# Uncomment the line above and comment the line below if you face an import error
from llama_index.embeddings.huggingface import HuggingFaceEmbedding

embed_model = HuggingFaceEmbedding(model_name="BAAI/bge-small-en-v1.5")
embeddings = embed_model.get_text_embedding("Hello World!")

print(len(embeddings))
print(embeddings[:5])

384
[-0.0032756798900663853, -0.011690747924149036, 0.041559211909770966, -0.03814811632037163, 0.02418307587504387]


Awesome! As you can see from the output above, we managed to download the `BAAI/bge-small-en-v1.5` model from Hugging Face by using Llama-Index and using the downloaded model to get the vector representation of the string "Hello World!".

### Let's Try a Few Embedding Models
As mentioned earlier in the quest, Hugging Face has many different models available for us to choose from. In this quest, we'll be using the following three models:

| model name                              | overall score | model params | model size | embedding length | url                                                            |
|-----------------------------------------|---------------|--------------|------------|------------------|----------------------------------------------------------------|
| BAAI/bge-small-en-v1.5                  | 62.x          | 33.5 M       | 133 MB     | 384              | https://huggingface.co/BAAI/bge-small-en-v1.5                  |
| sentence-transformers/all-mpnet-base-v2 | 57.8          |              | 438 MB     | 768              | https://huggingface.co/sentence-transformers/all-mpnet-base-v2 |
| sentence-transformers/all-MiniLM-L6-v2  | 56.x          |              | 91 MB      | 384              | https://huggingface.co/sentence-transformers/all-MiniLM-L6-v2  |

If you would like to, you can also download other models provided by [Hugging Face](https://huggingface.co/spaces/mteb/leaderboard) and compare the results you get from running them!

### Benchmark
What we're doing in the code cell below is running a benchmark test on our three chosen models. Essentially, we're downloading the models into our local machines, using them to create embeddings for the string "Hello World!" and comparing the speed of each model when it comes to generating embeddings.

Don't worry if you see the `TqdmWarning: IProgress not found` warning message, it should not affect your code from running.

After running the code cell below, if you look in the `cache` folder of the root directory of your project, you should be able to see that you've downloaded these three embedding models.

In [5]:
import time
import timeit

embedding_models = [
    'BAAI/bge-small-en-v1.5',
    'sentence-transformers/all-mpnet-base-v2',
    'sentence-transformers/all-MiniLM-L6-v2',
]

for model in embedding_models:
    embed_model = HuggingFaceEmbedding(model_name=model)
    embeddings = embed_model.get_text_embedding("Hello World!")
    print(f'model={model}, embeding_length={len(embeddings):,}')
    %timeit (embed_model.get_text_embedding("Hello World!"))
    print()

model=BAAI/bge-small-en-v1.5, embeding_length=384
20.1 ms ± 343 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)



modules.json: 100%|████████████████████████████████████████████████████████████████████| 349/349 [00:00<00:00, 296kB/s]
C:\Users\FaridahHusna\anaconda3\envs\atlas-1\Lib\site-packages\huggingface_hub\file_download.py:149: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\FaridahHusna\stackup-mongodb\cache\models--sentence-transformers--all-mpnet-base-v2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-de

model=sentence-transformers/all-mpnet-base-v2, embeding_length=768
42.7 ms ± 1.89 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)



modules.json: 100%|███████████████████████████████████████████████████████████████████████████| 349/349 [00:00<?, ?B/s]
C:\Users\FaridahHusna\anaconda3\envs\atlas-1\Lib\site-packages\huggingface_hub\file_download.py:149: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\FaridahHusna\stackup-mongodb\cache\models--sentence-transformers--all-MiniLM-L6-v2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to see activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-dev

model=sentence-transformers/all-MiniLM-L6-v2, embeding_length=384
10.8 ms ± 185 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)



## [OPTIONAL] Using SentenceTransformers
Note that besides using Llama-Index, you can also use other providers like SentenceTransformers to download embedding models; the code cell below demonstrates this.

In [1]:
from sentence_transformers import SentenceTransformer
# sentences = ["This is an example sentence", "Each sentence is converted"]

model = SentenceTransformer('sentence-transformers/all-MiniLM-L6-v2')
# embeddings = model.encode(sentences)
embeddings = model.encode('a happy dog!')
print(model)
print (len(embeddings))
print(embeddings[:5])

C:\Users\FaridahHusna\anaconda3\envs\atlas-1\Lib\site-packages\tqdm\auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


SentenceTransformer(
  (0): Transformer({'max_seq_length': 256, 'do_lower_case': False}) with Transformer model: BertModel 
  (1): Pooling({'word_embedding_dimension': 384, 'pooling_mode_cls_token': False, 'pooling_mode_mean_tokens': True, 'pooling_mode_max_tokens': False, 'pooling_mode_mean_sqrt_len_tokens': False, 'pooling_mode_weightedmean_tokens': False, 'pooling_mode_lasttoken': False, 'include_prompt': True})
  (2): Normalize()
)
384
[-0.00542401  0.07206918 -0.02727441  0.04371351 -0.06957791]


And we're done with this notebook! Please **head back to the Quest page on StackUp now**.